File: src/BootstrapBlazor/Components/Floating/Floating.razor.cs
// Copyright (c) Argo Zhang (argo@163.com). All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
// Website: https://www.blazor.zone or https://argozhang.github.io/

using Microsoft.AspNetCore.Components;

namespace BootstrapBlazor.Components;

/// <summary>
/// 浮动层 组件
/// </summary>
public sealed partial class Floating : IAsyncDisposable
{
    private FloatingContext? Context { get; set; }

    private FloatingState State { get; set; } = new();

    private readonly Offset _offset = new();

    /// <summary>
    /// 获得/设置 浮动层的锚点元素ID
    /// </summary>
    [Parameter]
    public string? Anchor { get; init; }

    /// <summary>
    /// 获得/设置 浮动层的锚点元素选择器
    /// </summary>
    [Parameter]
    public string? Selecter { get; init; }

    /// <summary>
    /// 获得/设置 浮动层的锚点组件
    /// </summary>
    [Parameter]
    public IdComponentBase? Component { get; init; }

    /// <summary>
    /// 获得/设置 浮动层的锚点元素
    /// </summary>
    [Parameter]
    public ElementWrapper? Wrapper { get; init; }

    /// <summary>
    /// 获得/设置 浮动层的锚点位置
    /// </summary>
    [Parameter]
    public Offset? Offset { get; init; }

    /// <summary>
    /// 获得/设置 浮动层显示状态
    /// </summary>
    [Parameter]
    public bool Visible
    {
        get => State.Visible;
        set
        {
            if (State.Visible != value)
            {
                State.Visible = value;
                ShouldCompute = true;
            }

            //判断是不是虚拟元素，且位置有没有发生变化
            if (value && CheckOffsetChanged())
            {
                ShouldCompute = true;
            }
        }
    }

    /// <summary>
    /// 获得/设置 浮动层相对于它的锚点元素的位置
    /// </summary>
    [Parameter]
    public Placement Placement
    {
        get => State.Placement;
        set
        {
            if (State.Placement != value)
            {
                State.Placement = value;
                if (Visible)
                {
                    ShouldCompute = true;
                }
            }
        }
    }

    /// <summary>
    /// 获得/设置 浮动层显示时 使用的定位模式，默认绝对定位
    /// </summary>
    [Parameter]
    public Position Position
    {
        get => State.Position;
        init => State.Position = value;
    }

    /// <summary>
    /// 获得/设置 内层子组件内容 
    /// </summary>
    [Parameter]
    public RenderFragment? ChildContent { get; init; }

    /// <summary>
    /// 获得/设置 浮动层渲染容器编号
    /// </summary>
    [Parameter]
    public string? ContainerId
    {
        get => State.ContainerId;
        init => State.ContainerId = value;
    }

    /// <summary>
    /// 获得/设置 当与锚点元素不在同一个剪辑上下文中时是否自动隐藏
    /// </summary>
    [Parameter]
    public bool AutoHide
    {
        get => State.AutoHide;
        init => State.AutoHide = value;
    }

    /// <summary>
    /// 获得/设置 是否自动更新位置，默认false
    /// </summary>
    [Parameter]
    public bool AutoUpdate
    {
        get => State.AutoUpdate;
        init => State.AutoUpdate = value;
    }

    /// <summary>
    /// 获得/设置 浮动元素和参考元素之间的距离。
    /// </summary>
    [Parameter]
    public int? AxisOffset
    {
        get => State.AxisOffset;
        init => State.AxisOffset = value;
    }

    /// <summary>
    /// 获得/设置 是否自动更改浮动元素的位置以使其处于可见状态。
    /// </summary>
    [Parameter]
    public bool UseFlip
    {
        get => State.UseFlip;
        init => State.UseFlip = value;
    }

    /// <summary>
    /// 获得/设置 配置需要填充的宽度，沿指定轴移动浮动元素以使其保持在视图中。
    /// </summary>
    [Parameter]
    public int? ShiftPadding
    {
        get => State.ShiftPadding;
        init => State.ShiftPadding = value;
    }

    /// <summary>
    /// 获得/设置 是否使用 提示箭头，浮动层必须预先设置为绝对定位
    /// </summary>
    [Parameter]
    public bool UseArrow
    {
        get => State.UseArrow;
        init => State.UseArrow = value;
    }

    /// <summary>
    /// 获得/设置 提示箭头的偏移量
    /// </summary>
    [Parameter]
    public int ArrowOffset
    {
        get => State.ArrowOffset;
        init => State.ArrowOffset = value;
    }

    /// <summary>
    /// 获得/设置 提示箭头的样式
    /// </summary>
    [Parameter]
    public string? ArrowStyleClass
    {
        get => State.ArrowStyleClass;
        init => State.ArrowStyleClass = value;
    }

    /// <summary>
    /// 获得/设置 组件样式类
    /// </summary>
    [Parameter]
    public string? @Class
    {
        get => State.Class;
        init => State.Class = value;
    }

    /// <summary>
    /// 获得/设置 组件样式
    /// </summary>
    [Parameter]
    public string? @Style
    {
        get => State.Style;
        init => State.Style = value;
    }

    /// <summary>
    /// 获得/设置 Floating 服务实例
    /// </summary>
    [Inject]
    [NotNull]
    private FloatingContainerService? Service { get; set; }

    /// <summary>
    /// 获得/设置 是否使用容器渲染
    /// </summary>
    private bool UseContainer { get; set; }

    /// <summary>
    /// 获得/设置 是否需要计算方位
    /// </summary>
    private bool ShouldCompute { get; set; }

    /// <summary>
    /// 获得/设置 是否已初始化渲染
    /// </summary>
    private bool FirstRendered { get; set; }

    /// <summary>
    /// OnInitialized
    /// </summary>
    protected override void OnInitialized()
    {
        base.OnInitialized();

        CheckParameter();

        UseContainer = Service?.ContainerHasRegistered(ContainerId) ?? false;

        Context = new FloatingContext(Id!, ChildContent, State);
        if (UseContainer)
        {
            Service?.RegisterContext(Context);
        }
    }

    /// <summary>
    /// OnParametersSetAsync
    /// </summary>
    /// <returns></returns>
    protected override async Task OnParametersSetAsync()
    {
        await base.OnParametersSetAsync();

        if (ShouldCompute)
        {
            ShouldCompute = false;

            //缓存虚拟元素位置，供比对使用
            if (State.Category == FloatingCategory.Virtual)
            {
                _offset.Left = Offset!.Left;
                _offset.Top = Offset!.Top;
                _offset.Width = Offset!.Width;
                _offset.Height = Offset!.Height;
            }

            if (FirstRendered)
            {
                await Toggle();
            }
            else
            {
                Context!.ExecuteAfterRender(Toggle);
            }
        }
    }

    /// <summary>
    /// OnAfterRenderAsync
    /// </summary>
    /// <param name="firstRender"></param>
    /// <returns></returns>
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            FirstRendered = true;
        }

        await base.OnAfterRenderAsync(firstRender);
    }

    /// <summary>
    /// ShouldRender
    /// </summary>
    /// <returns></returns>
    protected override bool ShouldRender() => false;

    /// <summary>
    /// 检查参数
    /// </summary>
    /// <exception cref="NotSupportedException"></exception>
    /// <exception cref="ArgumentNullException"></exception>
    private void CheckParameter()
    {
        if (Position != Position.Absolute && Position != Position.Fixed)
        {
            throw new NotSupportedException("Parameter Position：Only absolute and fixed are supported.");
        }

        if (!string.IsNullOrWhiteSpace(Anchor))
        {
            State.Category = FloatingCategory.Identifier;
        }
        else if (Wrapper != null)
        {
            State.Category = FloatingCategory.Element;
        }
        else if (Component != null)
        {
            State.Category = FloatingCategory.Component;
        }
        else if (!string.IsNullOrWhiteSpace(Selecter))
        {
            State.Category = FloatingCategory.Selecter;
        }
        else if (Offset != null)
        {
            State.Category = FloatingCategory.Virtual;
        }
        else
        {
            throw new ArgumentNullException(nameof(Anchor), "An anchor element must be set.");
        }
    }

    /// <summary>
    /// 检查锚点坐标是否发生变化
    /// </summary>
    private bool CheckOffsetChanged()
    {
        var isVitural = FirstRendered
            ? State.Category == FloatingCategory.Virtual
            : string.IsNullOrWhiteSpace(Anchor)
                && Wrapper == null
                && Component == null
                && string.IsNullOrWhiteSpace(Selecter)
                && Offset != null;

        if (isVitural)
        {
            return Offset!.Left != _offset.Left ||
                Offset!.Top != _offset.Top ||
                Offset!.Width != _offset.Width ||
                Offset!.Height != _offset.Height;
        }

        return false;
    }

    /// <summary>
    /// 切换浮动显示
    /// </summary>
    /// <returns></returns>
    private async ValueTask Toggle()
    {
        switch (State.Category)
        {
            case FloatingCategory.Identifier:
               await Compute(Anchor);
                break;
            case FloatingCategory.Element:
                await Compute(Wrapper?.Ref);
                break;
            case FloatingCategory.Component:
                await Compute(Component?.Id);
                break;
            case FloatingCategory.Selecter:
                await Compute(Selecter);
                break;
            case FloatingCategory.Virtual:
                await Compute(Offset);
                break;
        }
    }

    /// <summary>
    /// 计算浮动位置
    /// </summary>
    /// <typeparam name="TAnchor">元素类型</typeparam>
    /// <param name="anchor">锚点元素</param>
    /// <returns></returns>
    private async ValueTask Compute<TAnchor>(TAnchor anchor)
    {
        if (Context != null && anchor != null)
        {
            await JSRuntime.InvokeVoidAsync(Id, "bb_floating_toggle", anchor, Context.Element, State);
        }
    }

    /// <summary>
    /// Dispose
    /// </summary>
    /// <param name="disposing"></param>
    private async ValueTask DisposeAsyncCore(bool disposing)
    {
        if (disposing)
        {
            if (FirstRendered && AutoUpdate)
            {
                await JSRuntime.InvokeVoidAsync(Id, "bb_floating_cleanup");
            }
            if (UseContainer && Context != null)
            {
                Service?.UnregisterContext(Context);
            }
        }
    }

    /// <summary>
    /// Dispose 方法
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        await DisposeAsyncCore(true);
        GC.SuppressFinalize(this);
    }
}

File: src/BootstrapBlazor/Components/Floating/Floating.razor.js
(function ($) {
    $.extend({
        bb_floating_toggle: function (id, anchor, floating, option) {
            //移除自动浮动监听
            $.bb_floating_cleanup(id);

            //获取锚点元素
            let reference, indicator = null;
            switch (option.category) {
                case 1:
                case 3:
                    reference = document.getElementById(anchor);
                    break;
                case 2:
                    reference = anchor;
                    break;
                case 4:
                    reference = document.querySelector(anchor);
                    break;
                case 5:
                    reference = {
                        getBoundingClientRect() {
                            return {
                                width: anchor.width,
                                height: anchor.height,
                                x: anchor.left,
                                y: anchor.top,
                                top: anchor.top,
                                left: anchor.left,
                                right: anchor.left,
                                bottom: anchor.top,
                            };
                        },
                    };
                    break;
            }

            if (reference == null) {
                console.error("Reference not found.");
                return;
            }

            //配置中间件
            let config = { strategy: option.position, placement: option.placement, middleware: [] };

            //设置偏移量
            if (option.axisOffset) {
                config.middleware.push(FloatingUIDOM.offset(option.axisOffset));
            }

            //设置自动翻动来保持可见
            if (option.useFlip) {
                config.middleware.push(FloatingUIDOM.flip());
            }

            //设置保持内容完整显示
            if (option.shiftPadding) {
                config.middleware.push(FloatingUIDOM.shift({ padding: option.shiftPadding }));
            }

            //设置指示箭头
            if (option.useArrow) {
                indicator = floating.querySelector('[data-element=arrow]');
                if (indicator != null) {
                    config.middleware.push(FloatingUIDOM.arrow({ element: indicator }));
                }
            }

            //设置自动隐藏
            if (option.autoHide) {
                config.middleware.push(FloatingUIDOM.hide());
            }

            //控制显示状态
            if (option.visible) {
                if (floating.style.display == 'none') {
                    Object.assign(floating.style, {
                        display: 'block',
                        position: option.strategy,
                        left: null,
                        top: null,
                        opacity: 0,
                        transform: `scale(0)`,
                        transition: 'transform 0.2s ease, opacity 0.1s ease',
                        'z-index': -1000,
                    });
                }
            } else {
                if (floating.style.display != 'none') {
                    Object.assign(floating.style, { display: 'none' });
                }
                return;
            }

            //定义更新方法
            async function update() {
                //隐藏时不计算
                if (floating.style.display == 'none') {
                    return;
                }

                await FloatingUIDOM.computePosition(reference, floating, config)
                    .then(({ x, y, placement, strategy, middlewareData }) => {
                        const left = `${Math.round(x)}px`;
                        const top = `${Math.round(y)}px`;

                        if (floating.style.left == left && floating.style.top == top) {
                            return;
                        }

                        if (indicator != null) {
                            const p = placement.split('-')[0];
                            const side = { top: 'bottom', right: 'left', bottom: 'top', left: 'right' }[p] ?? 'bottom';
                            const offset = `${option.arrowOffset}px`;
                            let left1 = '', top1 = '';
                            if (middlewareData.arrow != undefined) {
                                if (middlewareData.arrow.x != undefined) {
                                    left1 = `${Math.round(middlewareData.arrow.x)}px`;
                                }
                                if (middlewareData.arrow.y != undefined) {
                                    top1 = `${Math.round(middlewareData.arrow.y)}px`;
                                }
                            }

                            Object.assign(indicator.style, { position: strategy, top: top1, left: left1, [side]: offset });
                        }

                        Object.assign(floating.style, {
                            position: strategy,
                            left: left,
                            top: top,
                            opacity: 1,
                            transform: `scale(1)`,
                            transition: 'transform 0.2s ease, opacity 0.1s ease',
                            'z-index': null,
                        });

                        if (option.autoHide) {
                            Object.assign(floating.style, {
                                visibility: middlewareData.hide.referenceHidden ? 'hidden' : 'visible'
                            });
                        }
                    });
            }

            //第一次需要手动调用
            update();

            //注册自动更新
            if (option.autoUpdate && option.category != 5) {
                window.floatings.set(id, FloatingUIDOM.autoUpdate(reference, floating, update));
            }
        },
        bb_floating_cleanup: function (id) {
            if (!window.floatings) {
                window.floatings = new Map();
            } else if (window.floatings.has(id)) {
                window.floatings.get(id).call(globalThis);
                window.floatings.delete(id);
            }
        }
    });
})(jQuery);
